### ISSUE TITLE
User installable Monterey plugins

### ISSUE THREAD

**Issue by adriatic** [monterey-framework/articles#1]
This article is referenced from **[Architecture improvements](https://github.com/monterey-framework/monterey/issues/424#issuecomment-273234331)** task
***

**Comment by adriatic** [comment]
Starting with the description of the Word Count plugin (aka package)  - defined **[here](http://flight-manual.atom.io/hacking-atom/sections/package-word-count/)**

**Comment by adriatic** [comment]
I am very sure that this is the way to go - Atom is incredibly popular, open source, maintained by GitHub, just like Electron (the basic Monterey infrastructure); just like it was all created for us.

**Comment by adriatic** [comment]
In order to prototype this concept, I will add it to the **[`Prototyper `](https://github.com/monterey-framework/prototyper)** app (what a weird concept 😏) by trying first to enable the **`Prototyper`**, to create a plugin using the **[`package generator`](https://github.com/atom/package-generator)**.

The second step would be to add the code that will allow the **`Prototyper`** to load and use the just created plugin.

**Comment by JeroenVinke** [comment]
While I would love to use Atom's plugin infrastructure, I do not believe that there is a platform abstraction layer that abstracts away atom's specific calls. For example, the package-generator plugin uses atom all over the place ([first](https://github.com/atom/package-generator/blob/master/lib/package-generator-view.coffee#L3), [second](https://github.com/atom/package-generator/blob/master/lib/package-generator-view.coffee#L43), [third](https://github.com/atom/package-generator/blob/master/lib/package-generator-view.coffee#L114) and more). In addition, Monterey does not have a code editor, so a large part of Atom plugins may not be applicable to Monterey. What do you think about this?

**Comment by adriatic** [comment]
I agree with your concerns (and have them as well). This is why I did not want to further elaborate on that before knowing that I brought you to this level of understanding. Here are the alternatives as I see them:

#### 1. Do nothing now
Driven by the desire to release Monterey.next as soon as possible after the new aurelia-cli with package importer, we leave the problem of using Atom's plugin infrastructure for next good opportunity. I wrote a message to Ken Sawicki about the platform abstraction layer you mentioned above - and am waiting for his answer; if he reacts positively we **should** wait for such layer to be created. More likely, we would be advised to do it ourselves, I think.

_In this situation, everything I describe in alternative 2 below, gets done on Monterey's own GUI._

#### 2. Use Atom to enhance Aurelia-CLI `Package importer`
We (at least I do) want to **[integrate Atom with Monterey](https://github.com/monterey-framework/monterey/issues/403)**. In this scenario, Atom could be used not only to edit the project's source code files, but also all configuration files - like package.json for example as shown below:

![image](https://cloud.githubusercontent.com/assets/2712405/22118758/03319424-de47-11e6-9997-afde223bafac.png 

on panel 3. Given the wealthy of information that the new `package importer` provides in the package importing process, having this interaction take place in Atom seems like a very appealing, user friendly process.

#### 3. Go full bore
Write Montery's own plugin infrastructure, strictly following Atom's so that all of the current Monterey's plugins are rewritten using this infrastructure. As stated ago, this would be incredibly valuable score for Monterey - but it would require two more people working on that.

**Comment by adriatic** [comment]
My comments to **[Jeroen's concerns stated above](https://github.com/monterey-framework/articles/issues/1#issuecomment-273825884)**:

I agree with the general problem of dependency with Atom editor (like **[this](https://github.com/atom/package-generator/blob/master/lib/package-generator-view.coffee#L3)**) - which is not an integrated component of Monterey. However, this is not the initial direction I would like to go. Instead, I would like to adopt the Atom's plugin format and implement our existing plugins using this format. Then I would also add the ** `Generate package`** component to Monterey

![image](https://cloud.githubusercontent.com/assets/2712405/22298944/4b2d8794-e2f0-11e6-85a7-457e57de7b7c.png)

which would allow Monterey users to create other Monterey packages (which would not have anything to do with Atom Editor, of course) - we will just use Atom compatible format.

**Comment by adriatic** [comment]
#### So far, not a peep from Kevin Sawicki - meaning that I need to up the effort to get his attention

**Comment by JeroenVinke** [comment]
> I agree with the general problem of dependency with Atom editor (like this) - which is not an integrated component of Monterey. However, this is not the initial direction I would like to go. Instead, I would like to adopt the Atom's plugin format and implement our existing plugins using this format. Then I would also add the ** Generate package** component to Monterey

Meaning that we adopt the plugin format, but existing atom packages are not supported by Monterey?

**Comment by adriatic** [comment]
> ... existing atom packages are not supported by Monterey?

Not until it makes sense to support them all - which would happen only when Atom is integrated with Monterey (___if ever 😄___)

There might me more than one existing package (like the package shown on the **[above screenshot](https://github.com/monterey-framework/articles/issues/1#issuecomment-275154402)** - **`package generator`** package.
